File: 10 Days of Statistics/Day 0: Weighted Mean.py
# challenge: https://www.hackerrank.com/challenges/s10-weighted-mean/problem

"""
Objective
In this challenge, we practice calculating a weighted mean.

Task
Given an array, X, of N integers and an array, W, representing the respective weights of X's elements, calculate and print the weighted mean of X's elements.
Your answer should be rounded to a scale of 1 decimal place.
"""

# numpy not allowed for this challenge
# import numpy as np gives ModuleNotFoundError: No module named 'numpy'

# the following is based on numpy that cannot be used
"""
def weighted_mean(ls_grade, ls_weight):
    mult = np.multiply(ls_grade, ls_weight)
    mult_sum = mult.sum()
    weight_sum = sum(ls_weight)
    return round(mult_sum/weight_sum,1)

weighted_mean(ls_grade, ls_weight)
"""

def weighted_mean(ls_grade, ls_weight):
    mult = [a * b for a, b in zip(ls_weight, ls_grade)]
    mult_sum = sum(mult)
    weight_sum = sum(ls_weight)
    return round(mult_sum/weight_sum,1)

# number of elements in list computed using len()
# following variable only needed for the challenge to work but not for the script
n=input()

ls_grade = list(map(int, input().strip().split()))

ls_weight = list(map(int, input().strip().split()))

print(weighted_mean(ls_grade, ls_weight))

File: 10 Days of Statistics/Day 0: Weighted Mean.R
# challenge: https://www.hackerrank.com/challenges/s10-weighted-mean/problem

con <- file("stdin", open = "r")

dataset <- readLines(con)

v_grade <- unlist(strsplit(dataset[2], split=" "))
v_grade <- as.numeric(v_grade)

v_weight <- unlist(strsplit(dataset[3], split=" "))
v_weight <- as.numeric(v_weight)

mult = v_grade*v_weight
mult_sum = sum(mult)
weight_sum = sum(v_weight)

cat(format(round(mult_sum/weight_sum, digits=1), nsmall=1))

close(con)
